_posts/2020-04-09-flink-serialization-tuning-vol-1.md

```diff
@@ -0,0 +1,277 @@
+```
+
+<div class="alert alert-warning" markdown="1">
+<span class="label label-warning" style="display: inline-block"><span class="glyphicon glyphicon-warning-sign" aria-hidden="true"></span> Warning</span>
+If you fail to provide the type information for a `Row`, Flink identifies that `Row` is not a valid POJO type according to the rules above and falls back to Kryo serialization (see below) which you will also see in the logs as:
+
+`13:10:11,148 INFO  org.apache.flink.api.java.typeutils.TypeExtractor             - Class class org.apache.flink.types.Row cannot be used as a POJO type because not all fields are valid POJO fields, and must be processed as GenericType. Please read the Flink documentation on "Data Types & Serialization" for details of the effect on performance.`
+</div>
+
+## Avro
+
+Flink offers built-in support for the [Apache Avro](http://avro.apache.org/) serialization framework (currently using version 1.8.2) by adding the `org.apache.flink:flink-avro` dependency into your job. Flink’s [AvroSerializer](https://github.com/apache/flink/blob/release-1.10.0/flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java) can then use Avro’s specific, generic, and reflective data serialization and make use of Avro’s performance and flexibility, especially in terms of [evolving the schema](https://avro.apache.org/docs/current/spec.html#Schema+Resolution) when the classes change over time.
+
+### Avro Specific
+
+Avro specific records will be automatically detected by checking that the given type’s type hierarchy contains the `SpecificRecordBase` class. You can either specify your concrete Avro type, or—if you want to be more generic and allow different types in your operator—use the `SpecificRecordBase` type (or a subtype) in your user functions, in `ResultTypeQueryable#getProducedType()`, or in `SingleOutputStreamOperator#returns()`. Since specific records use generated Java code, they are strongly typed and allow direct access to the fields via known getters and setters.
+
+<div class="alert alert-warning" markdown="1">
+<span class="label label-warning" style="display: inline-block"><span class="glyphicon glyphicon-warning-sign" aria-hidden="true"></span> Warning</span> If you specify the Flink type as `SpecificRecord` and not `SpecificRecordBase`, Flink will not see this as an Avro type. Instead, it will use Kryo to de/serialize any objects which may be considerably slower. 
+</div>
+
+### Avro Generic
+
+Avro’s `GenericRecord` types can, unfortunately, not be used automatically since they require the user to [specify a schema](https://avro.apache.org/docs/1.8.2/gettingstartedjava.html#Serializing+and+deserializing+without+code+generation) (either manually or by retrieving it from some schema registry). With that schema, you can provide the right type information by either of the following options just like for the Row Types above:
```

Comment:
```suggestion
Avro’s `GenericRecord` types cannot, unfortunately, be used automatically since they require the user to [specify a schema](https://avro.apache.org/docs/1.8.2/gettingstartedjava.html#Serializing+and+deserializing+without+code+generation) (either manually or by retrieving it from some schema registry). With that schema, you can provide the right type information by either of the following options just like for the Row Types above:
```